include/oneapi/dpl/internal/exclusive_scan_by_segment_impl.h

```diff
@@ -95,18 +96,52 @@ exclusive_scan_by_segment_impl(Policy&& policy, InputIterator1 first1, InputIter
 }
 
 #if _ONEDPL_BACKEND_SYCL
+
 template <typename Policy, typename InputIterator1, typename InputIterator2, typename OutputIterator, typename T,
           typename BinaryPredicate, typename Operator>
 oneapi::dpl::__internal::__enable_if_hetero_execution_policy<typename ::std::decay<Policy>::type, OutputIterator>
-exclusive_scan_by_segment_impl(Policy&& policy, InputIterator1 first1, InputIterator1 last1, InputIterator2 first2,
-                               OutputIterator result, T init, BinaryPredicate binary_pred, Operator binary_op)
+exclusive_scan_by_segment_impl_helper(Policy&& policy, InputIterator1 first1, InputIterator1 last1,
+                                      InputIterator2 first2, OutputIterator result, T init, BinaryPredicate binary_pred,
+                                      Operator binary_op, ::std::true_type /* has_known_identity*/)
 {
     const auto n = ::std::distance(first1, last1);
 
     // Check for empty element ranges
     if (n <= 0)
         return result;
 
+    typedef uint64_t CountType;
+
+    namespace __bknd = oneapi::dpl::__par_backend_hetero;
+
+    auto keep_keys = oneapi::dpl::__ranges::__get_sycl_range<__bknd::access_mode::read, InputIterator1>();
+    auto key_buf = keep_keys(first1, last1);
+    auto keep_values = oneapi::dpl::__ranges::__get_sycl_range<__bknd::access_mode::read, InputIterator2>();
+    auto value_buf = keep_values(first2, first2 + n);
+    auto keep_value_outputs = oneapi::dpl::__ranges::__get_sycl_range<__bknd::access_mode::write, OutputIterator>();
+    auto value_output_buf = keep_value_outputs(result, result + n);
+    auto buf_view = key_buf.all_view();
+    using iter_value_t = typename ::std::iterator_traits<InputIterator2>::value_type;
+
+    iter_value_t identity = __dpl_sycl::__known_identity<Operator, iter_value_t>::value;
+
+    sycl_scan_by_segment_impl<scan_type::exclusive> scan;
+
+    scan(::std::forward<Policy>(policy), key_buf.all_view(), value_buf.all_view(), value_output_buf.all_view(),
+         binary_pred, binary_op, init, identity);
+
+    return result + n;
+}
+
+template <typename Policy, typename InputIterator1, typename InputIterator2, typename OutputIterator, typename T,
+          typename BinaryPredicate, typename Operator>
+oneapi::dpl::__internal::__enable_if_hetero_execution_policy<typename ::std::decay<Policy>::type, OutputIterator>
+exclusive_scan_by_segment_impl_helper(Policy&& policy, InputIterator1 first1, InputIterator1 last1,
+                                      InputIterator2 first2, OutputIterator result, T init, BinaryPredicate binary_pred,
+                                      Operator binary_op, ::std::false_type /* has_known_identity*/)
+{
+
```

Comment:
Currently, in this PR, we check if the binary_op and type have a "known identity" according to SYCL, and only use the rewrite implementation if they have a "known identity".  If they do not, we use the previous implementation.  

It's come to my attention that there is a known limitation which was lost from the documentation (and will be reintroduced) which requires that the default constructor for the scan algorithms be the identity for the type and binary_op.    In practice, I believe it is only truly required in a certain combination of compiler and backend (due to a requirement of openmp simd scan).  


I think we have two options here: 
1) Always use the new implementation of the scan_by_segment algorithms for SYCL supported backend, and define the identity to be the default constructor wherever SYCL known identity does not exist.  When we reintroduce the known limitation to the documentation, keep the language broad.
2) Use the previous implementation for cases which do not have a "known identity". When we reintroduce the known limitation to the documentation, adjust the language of the specify the limitation only in the limited scope where it functionally exists in the implementation.


Option (1) will result in faster performance for cases which do not have a sycl known identity. However, it has the possibility to break some existing user code which worked in previous versions.  (2) Will keep the same level of support as we currently provide for binary_op / type combinations where the default constructor is not the identity.  However, the cases in question are likely rare end-cases.

Reply:
I think option 2 is the more conservative approach but is the current best option. For most use cases, I suppose that the sycl known identity will be present and the newer implementation would be invoked. Even if the cases in which existing user code could break are quite rare, I think this occurrence could cause quite some confusion since the limitation was not specified in the previous documentation.

Reply:
It's not a conservative. It's what we should do. Breaking the existing code is not an option. We (oneDPL) are generic library and we should care about our previous design and API even if we some internal optimization based on "something known :)" We could break something if we believe something is worth deprecating. But that's a separate discussion